sw/device/tests/rv_core_ibex_epmp_test.c

```diff
@@ -0,0 +1,336 @@
+      OT_UNREACHABLE();
+  };
+}
+
+/**
+ * Finds offset of a region's configuration in it's pmpcfg CSR.
+ *
+ * @param region PMP region.
+ * @return pmpcfg offset.
+ */
+inline uint32_t region_offset(uint32_t region) { return region % 4 * 8; }
+
+/**
+ * Sets up the execution area of Machine Mode.
+ *
+ * This configuration adjusts the existing configuration from the
+ * [reset PMP configuration](/hw/ip/rv_core_ibex/rtl/ibex_pmp_reset.svh)
+ * and [SRAM loader](/sw/host/opentitanlib/src/test_utils/load_sram_program.rs).
+ *
+ * These changes are needed before mseccfg.MML is enabled,
+ * because LRWX permissions become read only for machine mode.
+ */
+static void pmp_setup_machine_area(void) {
+  // Set up the writeable section of the SRAM executable.
+  //
+  // Note: this overlaps a region covering the whole of SRAM (region 15),
+  // but is in a lower PMP register so region 15's configuration
+  // will be ignored in this area.
+  const uint32_t rodata_end = (uint32_t)__rodata_end;
+  const uint32_t sram_end = TOP_EARLGREY_SRAM_CTRL_MAIN_RAM_BASE_ADDR +
+                            TOP_EARLGREY_SRAM_CTRL_MAIN_RAM_SIZE_BYTES;
+
+  CSR_WRITE(CSR_REG_PMPADDR8, tor_address(rodata_end));
+  CSR_WRITE(CSR_REG_PMPADDR9, tor_address(sram_end));
+
+  const uint32_t pmp9cfg = EPMP_CFG_A_TOR | EPMP_CFG_LRW;
+  CSR_SET_BITS(region_pmpcfg(9), pmp9cfg << region_offset(1));
+
+  // Clear the execution permissions on region 11 (MMIO)
+  CSR_CLEAR_BITS(region_pmpcfg(11), EPMP_CFG_X << region_offset(11));
+  uint32_t csr;
+  CSR_READ(region_pmpcfg(11), &csr);
+  CHECK(!((csr >> region_offset(11)) & EPMP_CFG_X),
+        "Couldn't remove execute access to PMP region 11.");
+
+  // Clear the write permissions on regions 13 (DV ROM) and 15 (all SRAM)
+  CSR_CLEAR_BITS(region_pmpcfg(13), EPMP_CFG_W << region_offset(13));
+  CSR_CLEAR_BITS(region_pmpcfg(15), EPMP_CFG_W << region_offset(15));
+  CSR_READ(region_pmpcfg(13), &csr);
+  CHECK(!((csr >> region_offset(13)) & EPMP_CFG_W),
+        "Couldn't remove write access from PMP region 15.");
+  CHECK(!((csr >> region_offset(15)) & EPMP_CFG_W),
+        "Couldn't remove write access from PMP region 15.");
+}
+
+/**
+ * Sets up the User Mode test function to be executable in user mode.
+ *
+ * *It wouldn't be very useful if it wasn't.*
+ */
+static void pmp_setup_user_area(void) {
+  const uintptr_t start = (uintptr_t)i_am_become_user;
+  const uintptr_t end = (uintptr_t)kIAmBecomeUserEnd;
+
+  CSR_WRITE(CSR_REG_PMPADDR0, tor_address(start));
+  CSR_WRITE(CSR_REG_PMPADDR1, tor_address(end));
+
+  const uint32_t pmp1cfg = (EPMP_CFG_A_TOR | EPMP_CFG_LRWX) ^ EPMP_CFG_R;
+  CSR_SET_BITS(region_pmpcfg(1), pmp1cfg << region_offset(1));
+}
+
+/**
+ * Sets up the permissions for the test locations.
+ *
+ * See the declaration of `test_locations`
+ * to see the desired permission settings.
+ */
+static void pmp_setup_test_locations(void) {
+  CSR_WRITE(CSR_REG_PMPADDR3, tor_address((uintptr_t)(test_locations + 0)));
```

Comment:
I think that this would make clearer that `test_locations` is an array and not arbitrary memory address.
```suggestion
CSR_WRITE(CSR_REG_PMPADDR3, tor_address((uintptr_t)(test_locations[0])));
```